include/taskolib/Sequence.h

```diff
@@ -107,21 +147,46 @@ class Sequence
     /// Alias for a constant reverse vector iterator.
     using ConstReverseIterator = std::vector<Step>::const_reverse_iterator;
 
-    /// Max number of bytes of a Sequence label.
+    /// Maximal number of bytes of a Sequence label.
     static constexpr std::size_t max_label_length = 128;
 
+    /// Default-construct an empty sequence with a random unique ID and an empty label.
+    Sequence();
```

Comment:
> Probably you want `Sequence` to be default constructable?

Yes, because why shouldn't it be? If I just want to construct an isolated `Sequence` in my program to execute it, I couldn't care less about a label, a unique ID, and a `SequenceManager`. In fact, almost all our unit tests are of that type.

> In principle, as the `Sequence` can not check if the ID is random or not, and it has all to be handled by the Sequence Manager anyhow, I would not allow this. A unique ID is needed, uniqueness has to be asserted from the outside.
> 
> What would I ever do with a Sequence with a Unique ID that is unchecked?
> 
> Also this connects uniqueness-checking (done presumably by the Sequence Manager) and ID generation to two unconnected entities; entwining the code of two objects with no obvious connection in the code.

I am not sure it is even possible to guarantee a "globally" unique ID. Take our own use case: We are running several Taskomat instances. Every one of them can of course guarantee uniqueness within its own sequence folder, but it does not know about the other instances. At some point, however, we want to link the sequence libraries via git, and there may be clashes then.

I think our best option is to 
1. make clashes of unique IDs very unlikely by default (e.g. by using random numbers)
2. add code that deals with clashes if they happen (e.g. by automatically assigning new unique IDs).

That is not to say that a virgin `Sequence` should not have a uniqueness check of some sort, but I am not sure how to do this elegantly. Specifically, I am not sure if we want to couple `Sequence` closely to `SequenceManager`. Too close coupling is the root of many evils.

Reply:
Well, I guess I would not couple `Sequence` back to `SequenceManager` at all.

Some ideas

* The `UniqueID` is 'dump', it just makes certain the ID (if we want to stick to strings) does not contain 'forbidden for filesystem' characters and also no `[` and `]`. This property enables us to add it to the directory name. Of course it could also be a 64b integer, that is then appended as hex or whatever.
* The `Sequence` does not care about its UniqueID, at least not more than about it's name. There must be one. It especially does not check it, why should it. Well, it must be 'creatable', which imposes the constraints mentioned above on it. But if the UniqueID is passed as UniqueID object when constructing a Sequence, there is nothing that needs to be done.
* At least technically a Sequence does not even need to KNOW its unique ID; maybe for display purposes for users that want to find it in the filesystem, but nothing else.
* The UniqueID needs to be unique in some context. A single Sequence is always 'unique'. Only collections of Sequences creates some environment that allows a definition of 'unique', i.e. there is no other with the same ID. If there is only one, that does not make sense. The context can be 'all Sequences stored/managed by this SequenceManager' or 'all Sequences in our git repo'.
* If Sequences are managed by a SequenceManager (and thus we have possibly more than one) the SequenceManager knows all the Sequences and the associated unique IDs that are living right now. It can create a new and guaranteed to be unused unique ID. For this it can either examine the local filesystem or query via git.
* This creates this flow of events
  * User wants to create a new Sequence. Taskomat asks SequenceManager to create a new (empty) Sequence. It is stored right away - unique ID is assigned and 'blocked'. Taskomat fills the new location with the (empty) Sequence by the normal readback procedure (I believe that is more or less already how it works).
  * User wants to copy a Sequence: This is handled via the SequenceManager anyhow. It just needs to copy the directory with a new unique ID appendix.

Maybe I overlooked something, but this seems simple enough.

The problem of merging different collections of Sequences should not be a taskolib problem. If there is no git and soneone wants to manually copy a sequence-directory from one account to another the person or a script can check the existing and new IDs and warn and/or automatically solve it. If it solves it - unfortunately - it needs some knowledge about the IDs. On the other hand we can say that a user needs to create an empty Sequence for each to-be-copied sequence (through the server via taskolib, this using the SequenceManager) and the script or admin just replaces the empty sequence with the to-be-copied sequence while keeping the empty sequence's ID. Maybe we need a 'fetch Sequence anew from disk' command in the server for this.
The same approach is possible for a git or sql based storage.

One other thing came to my mind ... If the unique ID has a fixed length, I would put that first in the directory name. The result is that the relative position of a sequence is fixed (i.e. 3rd from top on 'ls'), because the ID does not change. In the current situation the relative sorting will change when sequences names change. That might or might not be inconvenient.

`[0a64b811] My super duper sequence/`

Reply:
> One other thing came to my mind ... If the unique ID has a fixed length, I would put that first in the directory name. The result is that the relative position of a sequence is fixed (i.e. 3rd from top on 'ls'), because the ID does not change. In the current situation the relative sorting will change when sequences names change. That might or might not be inconvenient.
> 
> `[0a64b811] My super duper sequence/`

But most directory listings are sorted alphabetically, so this would make it very hard to find your sequence unless you memorized the hex number. This is why I proposed more mnemonic UIDs.

Reply:
I agree on most points. I would insist that a `Sequence` must be constructible in isolation, i.e. without a `SequenceManager`, but if I understand correctly you are not opposed to this.

This leaves the `SequenceManager` part to be implemented – so far, it does not have a function for creating new sequences and it does not really ensure uniqueness of the IDs. I'll work on this. If we should rather merge this MR as-is and add the other stuff later on, please let me know.

We disagree on how to migrate old sequences to the new format. I am too lazy to fiddle with shell scripts for the Debian packages, especially because the C++ code already solves the problem.

Reply:
> This leaves the `SequenceManager` part to be implemented – so far, it does not have a function for creating new sequences and it does not really ensure uniqueness of the IDs. I'll work on this. If we should rather merge this MR as-is and add the other stuff later on, please let me know.

I have now added member functions `create_sequence()` and `relabel_sequence()` to the `SequenceManager` class. Both `create_sequence()` and the conversion of legacy sequence folders ensure that their new UID is indeed unique in the sequences folder.